torch/csrc/jit/codegen/cuda/lower_trivial_broadcast.cpp

```diff
@@ -0,0 +1,176 @@
+#include <torch/csrc/jit/codegen/cuda/ir_utils.h>
+#include <torch/csrc/jit/codegen/cuda/iter_visitor.h>
+#include <torch/csrc/jit/codegen/cuda/lower2device.h>
+#include <torch/csrc/jit/codegen/cuda/root_domain_map.h>
+
+#include <torch/csrc/jit/codegen/cuda/lower_trivial_broadcast.h>
+
+namespace torch {
+namespace jit {
+namespace fuser {
+namespace cuda {
+
+void ConcretizedBroadcastDomains::build(Fusion* fusion) {
+  // Initialize the origin map with input broadcast domains
+  for (const auto fusion_input_tv :
+       ir_utils::filterByType<TensorView>(fusion->inputs())) {
+    for (auto root_id : fusion_input_tv->getRootDomain()) {
+      if (root_id->isBroadcast()) {
+        broadcast_origin_map_.emplace(
+            root_id, std::unordered_set<IterDomain*>({root_id}));
+      }
+    }
+  }
+  traverse(fusion);
+}
+
+bool ConcretizedBroadcastDomains::isConcretized(IterDomain* id) const {
+  auto it = concretized_domains_.find(id);
+  return it != concretized_domains_.end();
+}
+
+void ConcretizedBroadcastDomains::handle(BroadcastOp* bop) {
+  // Create a new entry for each of new broadcast domains
+  auto out = bop->out()->as<TensorView>();
+  for (const auto i : c10::irange(out->getRootDomain().size())) {
+    if (bop->getBroadcastDimFlags().at(i)) {
+      auto new_bcast_id = out->getRootDomain().at(i);
+      broadcast_origin_map_.emplace(
+          new_bcast_id, std::unordered_set<IterDomain*>({new_bcast_id}));
+    }
+  }
+}
+
+void ConcretizedBroadcastDomains::handle(Expr* expr) {
+  IterVisitor::handle(expr);
+
+  // Propagate broadcast origin info from producers to consumers
+  for (auto producer : ir_utils::filterByType<TensorView>(expr->inputs())) {
+    std::unordered_set<IterDomain*> producer_broadcasts;
+    for (auto producer_id : producer->getMaybeRFactorDomain()) {
```

Comment:
Can you mention this assumes broadcast axes are not merged ~with non-broadcast axes~ between root and rfactor?

Reply:
Added a comment.